### ISSUE TITLE
libwebsocket_callback_on_writable in multithreaded environment

### ISSUE THREAD

**Issue by heinemml** [warmcat/libwebsockets#314]
Hi,

I guess I'm picking up where Jeffrey Zhao left on the Mailinglist in April.

I have a multithreaded C++11 program and got the problem that libwebsocket_callback_on_writable doesn't trigger a LWS_CALLBACK_SERVER_WRITEABLE callback when called from a different thread than the one where the libwebsocket_service() call is issued.

The libwebsocket_service call is issued in a seperate thread which was created using std::thread. When I'm issuing libwebsocket_callback_on_writable from the callback itself the LWS_CALLBACK_SERVER_WRITEABLE callback is succesfully called.
When I'm doing the same thing from the main thread it is not.

What did I check:
  The wsi and wsContext pointers are pointing to the exact same address.
  When I'm checking for libwebsocket_get_socket_fd(wsi) I get the identical fd in both threads.

  I've implemented mutex locking in LWS_CALLBACK_LOCK_POLL and unlocking in LWS_CALLBACK_UNLOCK_POLL.
  LWS_CALLBACK_GET_THREAD_ID is returning the respective thread id.

My enviroment ist OS X 10.10.5 with libwebsockets 1.4 installed from homebrew. Compilation is done with clang using these flags: "-std=c++14 -O3 -Wall -Werror -pedantic -g"

Any ideas or hints?

Cheers,
Michael

**Comment by heinemml** [comment]
I investigated a bit further.
When compiling and running the exact same code on Linux it works as expected. I can send from both threads.

So I started looking around in the libwebsocket source. My current assumption is, that poll() does something different on OS X than on Linux. 

My way of thinking:
libwebsocket_callback_on_writable calls lws_change_pollfd(wsi, 0, LWS_POLLOUT). This sets fd.events of my file descriptor to POLLOUT and calls libwebsocket_cancel_service. I can verify this by printing out the fd.events value at the end of the function. It shows 5. Which is (POLLIN | POLLOUT)

I now print out "events" after the poll() function call returns in lws_plat_service(). This still prints out 1. So it won't trigger the write callback. So I'm suspecting poll() on OS X is somehow resetting the fds.event memory on returning. 

Why is it working from the loop thread? Beacause when I'm calling libwebsocket_callback_on_writable I'm not in poll() on any thread. So this would make sense.

To verify I'm working on the same memory I added a counter variable in the libwebsocket_context. Which can perfectly be increased in one thread and can be printed out correctly in both threads. 


Any ideas?

**Comment by heinemml** [comment]
Okay, I found a workaround which involves patching libwebsockets. Any comments on that?

````C
diff --git a/lib/lws-plat-unix.c b/lib/lws-plat-unix.c
index a7f96a3..a8bb06d 100644
--- a/lib/lws-plat-unix.c
+++ b/lib/lws-plat-unix.c
@@ -119,7 +119,15 @@ lws_plat_service(struct libwebsocket_context *context, int timeout_ms)
        if (lws_ssl_anybody_has_buffered_read(context))
                timeout_ms = 0;
 #endif
-       n = poll(context->fds, context->fds_count, timeout_ms);
+
+       struct libwebsocket_pollfd* fds_copy = malloc(sizeof(struct libwebsocket_pollfd) * context->fds_count);
+       memcpy(fds_copy, context->fds, sizeof(struct libwebsocket_pollfd)*context->fds_count);
+       n = poll(fds_copy, context->fds_count, timeout_ms);
+       for (int i = 0; i < context->fds_count; i++) {
+               context->fds[i].revents = fds_copy[i].revents;
+       }
+       free(fds_copy);
+
        context->service_tid = 0;
 
 #ifdef LWS_OPENSSL_SUPPORT
````

**Comment by heinemml** [comment]
I posted an example program demonstrating the difference in behavior on [Stack Overflow](http://stackoverflow.com/questions/32192063/differences-in-poll-between-linux-and-os-x)

**Comment by warmcat** [comment]
I don't understand your workaround. The whole idea of poll() is to leave a result in .revents.

**Comment by heinemml** [comment]
Okay, you didn't understand my workaround. But did you understand the problem? Let me reiterate:

When you call libwebsockets_callback_on_writeable() on a connection it basically sets the fds->events field here: https://github.com/warmcat/libwebsockets/blob/3ae1badae7a05e0982e0dfbcb078da3d4b92a81d/lib/pollfd.c#L150

If you're are multithreaded the service loop might currently be in poll() using the same fds.

So you are essentially modifying fds->events while it is used by loop. On Linux this seems to be fine(*) as there are no problems occurring. On OS X poll() seems to set fds->events back to the values it had when you're entering the function, basically overwriting any change you did to it while loop() was blocking. This leads to libwebsockets_callback_on_writeable() being essentially without effect I you're not making sure it is called while poll() is not currently running.

(*) From the responses on [Stack Overflow](http://stackoverflow.com/questions/32192063/differences-in-poll-between-linux-and-os-x) I get that is also not really fine because it is undefined behaviour and might or might not work.

**Comment by warmcat** [comment]
Well... your workaround is kind of far from the problem.  I'm glad I didn't understand it ^^

You shouldn't call the service loop from different threads.  Trying to 'work around' that is just chasing your own tail.

Have one thread deal with lws and the others signal him, or put stuff in a fifo for him, etc.

**Comment by heinemml** [comment]
Maybe I was not clear on that point:

I have one thread which loops and is only calling libwebsocket_service(wsContext, 100). If LWS_CALLBACK_SERVER_WRITEABLE occurs it reads stuff from the ringbuffer and sends it.
The other thread is putting stuff in the ringbuffer and then calls libwebsocket_callback_on_writable().

I have the exact same code running on Linux without problems and with huge delays/packet loss on OS X.

Only workaround without fiddling in libwebsockets directly: call libwebsocket_callback_on_writable() after each LWS_CALLBACK_SERVER_WRITEABLE. But this leaves you with 100% CPU Load because poll is returning instantly as you already set the event.

**Comment by heinemml** [comment]
Okay, to eliminate all "you don't use multithreading correctly" assumptions I put together a simple example:

main.cpp:
```c++
#include <libwebsockets.h>
#include <iostream>
#include <thread>
#include <atomic>

std::atomic<bool> running(true);
std::atomic<bool> connected(false);
struct libwebsocket_context *wsContext = nullptr;

static int mycallback(struct libwebsocket_context *context, struct libwebsocket *wsi, enum libwebsocket_callback_reasons reason, void *user, void *in, size_t len)
{
	switch (reason)
	{
		case LWS_CALLBACK_ESTABLISHED:
		{
			std::cout << "service: client connected" << std::endl;
			connected = true;
			break;
		}
		case LWS_CALLBACK_SERVER_WRITEABLE:
		{
			std::cout << "service: LWS_CALLBACK_SERVER_WRITEABLE" << std::endl;
			break;
		}
		case LWS_CALLBACK_GET_THREAD_ID:
		{
#ifdef __APPLE__		
	        uint64_t tid;
	        pthread_threadid_np(NULL, &tid);
			return tid;
#else
			return pthread_self();
#endif
		}
		default:
		{
			break;
		}
	}

	return 0;
}

struct libwebsocket_protocols protocols[] = {
    {"dummy", mycallback, 0, 100},
    {nullptr, nullptr, 0, 0}};

static void service_loop()
{
	std::cout << "service: thread starting" << std::endl;
	while (running)
	{
		libwebsocket_service(wsContext, 100000);
		std::cout << "service: returned" << std::endl;
	}
}

int main()
{
	struct lws_context_creation_info info;

	memset(&info, 0, sizeof info);
	info.port = 4711;
	info.iface = nullptr;
	info.protocols = protocols;
	info.ssl_cert_filepath = nullptr;
	info.ssl_private_key_filepath = nullptr;
	info.gid = -1;
	info.uid = -1;
	info.options = 0;

	wsContext = libwebsocket_create_context(&info);

	std::cout << "Websocket listening on port " << info.port << std::endl;

	std::thread service_thread(service_loop);

	while(!connected)
	{
		usleep(1000);	
	}
	
	std::cout << "main: schedule LWS_CALLBACK_SERVER_WRITEABLE" << std::endl;
	libwebsocket_callback_on_writable_all_protocol(&protocols[0]);
	sleep(1);
	std::cout << "main: schedule LWS_CALLBACK_SERVER_WRITEABLE" << std::endl;
	libwebsocket_callback_on_writable_all_protocol(&protocols[0]);
	sleep(1);
	std::cout << "main: schedule LWS_CALLBACK_SERVER_WRITEABLE" << std::endl;
	libwebsocket_callback_on_writable_all_protocol(&protocols[0]);

	sleep(5);

	running = false;
	libwebsocket_cancel_service(wsContext);

	service_thread.join();

	return 0;
}
```

Compiled with: 
```bash
c++ -std=c++14 -O3 -Wall -Werror -pedantic -g -I/usr/local/include    -o main.cpp.o -c main.cpp
c++ -std=c++14 -O3 -Wall -Werror -pedantic -g main.cpp.o  -o wsdemo  -L/usr/local/lib  -lwebsockets -lpthread
```

Running the program and connecting with a websocket client to port 4711 yields the following result:


Linux (Xubuntu 15.04)
```
./wsdemo 
[1444810036:4434] NOTICE: Initial logging level 7
[1444810036:4435] NOTICE: Library version: 1.4 f54715b
[1444810036:4435] NOTICE: IPV6 not compiled in
[1444810036:4435] NOTICE: libev support not compiled in
[1444810036:4435] NOTICE:  static allocation: 4536 + (16 x 1024 fds) = 20920 bytes
[1444810036:4436] NOTICE:  canonical_hostname = osboxes
[1444810036:4436] NOTICE:  per-conn mem: 248 + 2140 headers + protocol rx buf
[1444810036:4436] NOTICE:  Compiled with OpenSSL support
[1444810036:4436] NOTICE:  Using non-SSL mode
[1444810036:4451] NOTICE:  Listening on port 4711
Websocket listening on port 4711
service: thread starting
service: returned
service: client connected
service: returned
main: schedule LWS_CALLBACK_SERVER_WRITEABLE
service: returned
service: LWS_CALLBACK_SERVER_WRITEABLE
service: returned
main: schedule LWS_CALLBACK_SERVER_WRITEABLE
service: returned
service: LWS_CALLBACK_SERVER_WRITEABLE
service: returned
main: schedule LWS_CALLBACK_SERVER_WRITEABLE
service: returned
service: LWS_CALLBACK_SERVER_WRITEABLE
service: returned
service: returned
```

Mac (OS X 10.11)
```
./wsdemo
[1444810260:2640] NOTICE: Initial logging level 7
[1444810260:2641] NOTICE: Library version: 1.4 f54715b
[1444810260:2641] NOTICE: IPV6 not compiled in
[1444810260:2641] NOTICE: libev support not compiled in
[1444810260:2641] NOTICE:  static allocation: 4488 + (16 x 256 fds) = 8584 bytes
[1444810260:2641] NOTICE:  canonical_hostname = mheinemac.internal.novero.com
[1444810260:2641] NOTICE:  per-conn mem: 160 + 2140 headers + protocol rx buf
[1444810260:2658] NOTICE:  Listening on port 4711
Websocket listening on port 4711
service: thread starting
service: returned
service: client connected
service: returned
main: schedule LWS_CALLBACK_SERVER_WRITEABLE
service: returned
main: schedule LWS_CALLBACK_SERVER_WRITEABLE
service: returned
main: schedule LWS_CALLBACK_SERVER_WRITEABLE
service: returned
service: returned
```

As you can see the LWS_CALLBACK_SERVER_WRITEABLE is scheduled but never called. 
Maybe the problem is clearer now.

**Comment by JoakimSoderberg** [comment]
@heinemml  Not to tell you not to use libwebsockets. But since you're obviously a "C++ on the bleeding edge" kind of guy, how come you're not using something like https://github.com/zaphoyd/websocketpp with native C++ support? I'm just curious in general, because I've seen multiple C++ projects wrapping libwebsockets. Is it because the lightweight nature of libwebsockets? :)

**Comment by heinemml** [comment]
@JoakimSoderberg Valid point. It's more for historical reason. We started with some parts being in C and decided to stay consistent across the project.

**Comment by Grantley** [comment]
@heinemml Dude your a genius... I am having the exact same problem, could not work out the issue - thank you for your insight...

**Comment by heinemml** [comment]
@Grantley happy to hear that this post was helpful to someone. How did you work around this issue?

**Comment by RouquinBlanc** [comment]
Hi guys,

I'm seeing this issue as well on osX High Sierra with a fresh brew install of libwebsockets.

Modified code from @heinemml to work with lws 2.4.1: [pastbin](https://pastebin.com/KzHj8LM3)

This will output:
```
[2017/11/10 18:29:36:9207] NOTICE: libuv support not compiled in
[2017/11/10 18:29:36:9208] NOTICE: Creating Vhost 'default' port 4711, 1 protocols, IPv6 off
Websocket listening on port 4711
service: thread starting
service: returned
service: client connected
service: returned
service: returned
service: LWS_CALLBACK_SERVER_WRITEABLE
service: returned
main: schedule LWS_CALLBACK_SERVER_WRITEABLE
service: returned
main: schedule LWS_CALLBACK_SERVER_WRITEABLE
service: returned
main: schedule LWS_CALLBACK_SERVER_WRITEABLE
service: returned
service: returned
```

Not sure on what to do about:

- The fact that libuv support is not compiled in
- That initial call to `LWS_CALLBACK_SERVER_WRITEABLE` before any call to `lws_callback_on_writable(client_wsi)` ?

This is probably the reason behind my issue on janus gateway on OSX when using websockets: https://groups.google.com/forum/#!topic/meetecho-janus/HsFaEXBz4Cg

**Comment by lws-team** [comment]
No offense but finding meaning in 'necro bug shopping' like this

> There is an issue raised in 2015 (!!!) against libwebsockets which did not seem to reach any conclusion: https://github.com/warmcat/libwebsockets/issues/314

> I posted an update on that issue, let's see how this gets...

... instead of just describing YOUR problem without assumptions is not very useful; there's no action I can take about your assumption.  Lws has been going since 2010 (!!!) you can find bug reports of that vintage that have nothing to do with anything.

If you think I should spend my time on your problem, open a fresh issue explaining clearly why you think it's my problem and what action you think I should take about it.

**Comment by RouquinBlanc** [comment]
Hi Andy,

No Offense taken.
To make it simple, "my" problem happens when trying to run janus-gateway, whose websocket plugin is built on top of libwebsockets, is not working on mac os. To be more precise, when calling `lws_callback_on_writable` from the main thread after posting a message to be transmitted in the websocket callback upon LWS_CALLBACK_SERVER_WRITEABLE event, no call to the callback is triggered. This works perfectly in linux and we have been relying on it for quite a while already.

Now, I've been looking around, and this issue's description matches almost completely the situation I'm seeing, except that the APIs evolved quite a bit in between, but still the main problem remains: on mac os, calling the `lws_callback_on_writable` from another thread to wake up the lws thread does not seem to work, as the simple snippet from @heinemml is (still) showing.

I'm just trying to find a solution here, or at least the root cause of that behaviour and a way to work around it. Note that there are quite a few discussions about the differences of `poll` implementation on osx compared to Linux, which is the underlying implementation in lws AFAICS. [https://github.com/curl/curl/issues/1057](Libcurl) seems to have problems as well with it, even differences across versions.

I can open a new issue or discussion anywhere else if you think its more appropriate.

**Comment by lws-team** [comment]
"waking up poll() from another thread" actually means crafting an event on something the poll() is already listening to, there is no other way to interrupt it.

In lws on *nix, we have a private pipe added to the poll(), we can abort the poll() at any time from another thread by sending a byte into the other end of the pipe.  There is an api to do that, `lws_cancel_service()`.  On master, there is now a callback broadcase `LWS_CALLBACK_EVENT_WAIT_CANCELLED` that you only get if another thread woke you up by the `lws_cancel_service()` api.

If you implement `LWS_CALLBACK_GET_THREAD_ID` in protocols[0] to return a current thread ID int, lws will detect that you are calling `lws_callback_on_writable()` from a different thread and call `lws_cancel_service()` for you, to wake you up out of the poll.  On pre-master lws, you won't get the `LWS_CALLBACK_EVENT_WAIT_CANCELLED` broadcast though, just the callbacks when writable.

However, all that is just an optimization, because poll() has a timeout.  If your poll() was completely idle, it should wait through the timeout, and go back around the event loop to enter poll() again.  When that happens, it will pay attention to the new set of events to listen for, which should have included the `lws_callback_on_writable()` ones.  If that isn't happening just on mac platform, and with everything the same does work on linux platform, you are going to have to debug where the problem is coming from.... I have linux but I don't have (or want) a mac.

It does not sound like the problem in curl you mentioned, they say "There may seem a bug or change in the poll() syscall on Mac OS 12 that causes it to return prematurely.".  If all else was OK and poll() returned prematurely, we would just pick up the new events to poll() quicker and all would be well.

**Comment by RouquinBlanc** [comment]
Thanks a lot for the update.

The point on the Libcurl link is: poll is or can be acting differently on mac.

Even with a short timeout on the poll, when it loops it doesn't take the new event.

With the master branch, indeed I get the LWS_CALLBACK_EVENT_WAIT_CANCELLED event triggered when calling lws_cancel_service(), or even without calling it if, as suggested, I implement the `LWS_CALLBACK_GET_THREAD_ID` callback. But as you say, it's an optimisation. At the moment the timeout is set to 50ms, so only gain that much time (the timeout could be increased though with that new feature).

Now, it still does not trigger a writable event by itself. I could call  `lws_callback_on_writable_all_protocol` during the handling of cancelled event to force a wake up on all sockets, but that's not efficient...

**Comment by lws-team** [comment]
> Even with a short timeout on the poll, when it loops it doesn't take the new event.

Right, so it's not to do with poll() returning early, which is the curl bug.  Your problem is no matter how long you waited, there doesn't seem to be a POLLOUT coming for that wsi, even if you know it's writeable.

> Now, it still does not trigger a writable event by itself. 

There still seems something broken somewhere for that platform if I understood you... it should have set POLLOUT event active.  The best way is follow what happens when your other thread calls `lws_callback_on_writable()` and understand what's broken and fix or workaround it.

If you just want a workaround and can use master lws, then you can look at keeping a queue of wsi outside of lws that need writing, and go through calling `lws_callback_on_writable()` on them in the WAIT_CANCELLED callback.  But actually fixing it will be better for you too if you can trace what happens or doesn't happen when the other thread calls `lws_callback_on_writable()`.

**Comment by RouquinBlanc** [comment]
Hi Andy,

A fix would be indeed preferable for everyone.

Tracing a bit on what happens, the assumption of @heinemml seems correct. Added some traces in pollfd.c when updating the events, and just before and after `poll()`:

```
[2017/11/11 10:55:34:6861] NOTICE: [0x7fd821c02b48] before poll events:
[2017/11/11 10:55:34:6861] NOTICE: 	 fds[0] = 1
[2017/11/11 10:55:34:6861] NOTICE: 	 fds[1] = 1
[2017/11/11 10:55:34:6862] NOTICE: 	 fds[2] = 1
main: call lws_callback_on_writable(client_wsi)
service: 35 ()
[2017/11/11 10:55:34:6862] NOTICE: [0x7fd821c02b48] changing fds[2] from 1 to 5
service: 34 ()
service: 31 (LWS_CALLBACK_GET_THREAD_ID)
service: 36 ()
[2017/11/11 10:55:34:6863] NOTICE: [0x7fd821c02b48] after poll events:
[2017/11/11 10:55:34:6863] NOTICE: 	 fds[0] = 1
[2017/11/11 10:55:34:6863] NOTICE: 	 fds[1] = 1
[2017/11/11 10:55:34:6863] NOTICE: 	 fds[2] = 1
```

As discussed in the [stack overflow](https://stackoverflow.com/questions/32192063/differences-in-poll-between-linux-and-os-x-when-pollfd-is-changed-on-another-t) mentioned higher, it does not seem to be specified what poll should do about the `events` themselves, just about the `revents`. Indeed I don't see any mention of that as well in the BSD man page, and if you take a quick look at the [BSD poll implementation](https://github.com/apple/darwin-xnu/blob/0a798f6738bc1db01281fc08ae024145e84df927/bsd/kern/sys_generic.c#L1818), they indeed seem to copy the fds in poll, work on the copyand set the `revents`, then copy back the updated fds... It does not seem that far fetched as an implementation, and other platforms may well behave the same way (especially if copied from BSD!)

I would personally think that, to be safe, having `poll` working on a copy of the fds while other threads modify the context->fds events seems to be a safer approach, don't you think? And as such, I support the fix proposed [earlier](https://github.com/warmcat/libwebsockets/issues/314#issuecomment-134267891) in this thread, or some equivalent mechanism.

**Comment by lws-team** [comment]
> I would personally think that, to be safe, having poll working on a copy of the fds while other threads modify the context->fds events seems to be a safer approach, don't you think?

Sorry I don't understand what you think is happening... you think that before returning from poll(), the implementation writes to the userland buffer that contains the .events member?

There are no such shenanigans going on in Linux, so, no, copying around .events doesn't make anything 'safer'.  If you want to try doing that and seeing if it solves your problem (it shouldn't be that hard, search for poll() in lws-plat-unix.c and memcpy the pollfd array to a copy before giving it to poll, and copy back out the revents), we can discuss what to do about it if you find that magically makes your problem go away.  But reading BSD poll() manpage

https://www.freebsd.org/cgi/man.cgi?query=poll&sektion=2&apropos=0

it says nothing about changes to .events.  I guess a buttload of code would blow up if poll() started trashing .events.

**Comment by RouquinBlanc** [comment]
Look at the code: I'm not saying that "I think" that it writes to the land buffer: it just does.

And where do you think those traces are coming from? I did already do what you suggest, and the copy does fix the issue, both in the simple snippet provided and in the Janus library (provided that I use master and define a thread ID callback in protocols[0] as suggested).

Now generally, if the man page doesn't say if a function does or doesn't do something, l would assume it to be unspecified behaviour, and as such prone to implementation differences. Making a copy is safer because it does not depend on what the implementation is, safer in terms of cross platform behaviour, but also in case that the implementation in Linux changes (which is quite unlikely as it would break many thinks, as you said).

Just trying to help here! Please let me know if you want more tests from me on osx, would be glad to participate.

**Comment by lws-team** [comment]
Oh... well I guess it's good news I am not yet retarded enough to imagine BSD would do that; I am sorry I didn't get your point.  I take it you mean this, it just takes a dump on the whole userland struct

```
	error = copyout(fds, uap->fds, nfds * sizeof(struct pollfd));
```

In that case it's still not "playing it safe" (which gave me the wrong idea) but means BSD simply doesn't support other threads changing .events while poll() is active.

In the worst case, when entering poll() you may have to copy around a huge amount of struct pollfds if the solution is this naive "give the kernel a copy of all the pollfds".  I don't think that will fly.

What we could do instead is keep a list of pollfds modified while the service thread is in the poll wait, and apply the modifications on the original pollfds just before reentering the poll wait.

Modifications to pollfds *after* we are out of the poll() wait are unaffected... that is **all** modifications that happen from the service thread itself, since by definition it's not waiting in poll() if it's doing something that changes someone's .event.  So it's only guys from other threads that mess with .events while the main thread is in poll() that will get overwritten.

**Comment by lws-team** [comment]
Rereading what I wrote it doesn't carry what I meant very well... I really am sorry that I didn't get your point since you made it clearly enough.  But I am also kind of glad I didn't imagine BSD would do something like that since it's pretty broken.

I'll take a closer look at it tomorrow.  In theory there should only be one "other thread" pollfd change possible for each poll wait, since as part of its action it should force the poll() wait to end.  But I am worried this may require some kind of locking.

**Comment by RouquinBlanc** [comment]
Again, no problem. Just trying to make it all work for everyone, even the evil mac users ;-)

You are right, the copy/paste of fds does not scale very well (especially for `poll`, which unlike `select` is not limited to 1023 fds...). But yes, if you want to avoid useless copies of big structures, or lookup for flags on all the registered wsi, you may need a shared structure for each "other thread" (you could assume that in worst case, a client could spawn one thread per socket, all very actively writing out data), so that each of them just add themselves in that structure (queue?) when they want to modify their events, potentially at the same time... so shared == some kind of locking.

Honestly I'm a bit shared between on one side the copy of twice 8 bytes per active client websocket and on the other side the added cost and complexity of a locking system. I guess it all depends on how much complexity that would add! Nowadays copies are cheap, you could even use the stack up to a certain number of clients to avoid malloc/free.

Other solution (just for the sake of discussion): you said that when another thread calls `lws_callback_on_writable`, provided that the thread can be identified then `lws_cancel_service` is called, which if I understand well just puts a byte in a fifo to wake up the `poll`. What if instead of putting a byte, you would put information on which wsi triggered the cancellation (when suited), so that in turns when you dequeue that info you would know exactly and almost for free which wsi to update? Not sure if that makes sense or not 🍺 

Cheers
Jonathan

**Comment by lws-team** [comment]
I just made some coffee and sat down to look at this... using the cancel pipe is a neat idea if, eg, multibyte writes in there are atomic between threads.  But I think the main point is the locking.

There are also scenarios like the other thread told us to do RX flow control (same problem but with POLLIN) but a wsi getting serviced "naturally" countermanded it.  So we have a restriction we have to apply any pending business from the other thread first before servicing poll() events.

If we don't have locking, we can get a race that the other thread checks to see if he's in poll(), and he is, so he adds a packet in the pipe.  But while he did that, we left poll() and didn't see anything in the pipe yet, so we started normal servicing of events.  Then the other thread's request stays in the pipe until we get back to poll(), potentially now with an out-of-date change that the natural service countermanded already.

**Comment by lws-team** [comment]
I pushed a patch on master aimed at supporting this cheaply and automatically, but it's only build-tested.  Please give it a try (and help debug it if needed ;-) )

**Comment by RouquinBlanc** [comment]
I just had a look.

At first I doubted the `foreign_spinlock` efficiency when multiple clients, but then I noticed the that pollfd was modified under context mutex lock, so that's safe. Then on the main thread, the spin lock should protect enough because of the combination with volatile `inside_poll` 👍 

But still, with this simple [test code](https://pastebin.com/Bc8zZ8LJ), I end up either with a core dump or an infinite loop with head of `foreign_pfd_list` referencing itself:

```
104			while (*ftp1)  # pollfd.c:104
(gdb) p ftp1
$1 = (struct lws_foreign_thread_pollfd **) 0x100102380
(gdb) p *ftp1
$2 = (struct lws_foreign_thread_pollfd *) 0x100102380
(gdb) p **ftp1
$4 = {next = 0x100102380, fd_index = 2, _and = 0, _or = 4}
(gdb) 
```

Note that If I clear the list pointer before handling the different queued events, the issue disappears:

```c
	/* Collision will be extremely rare.  Just spin until it completes */
	while (pt->foreign_spinlock)
		;
	ftp = pt->foreign_pfd_list;
	pt->foreign_pfd_list = NULL;
	while (ftp) {
```
But isn't that hiding a problem?

**Comment by lws-team** [comment]
The spinlock is only for the case that when we come out of the poll wait(), which is something we can't control the timing of, we happen to also be making a pollfd change from the foreign loop.  So I think this will happen at a low probability, since we don't spend much time in the code making the pollfd change from the foreign loop, and coming out of the poll() wait is not something that happens that often.  So it is fine AFAIK.

Ehhh how can the ftp->next be itself?  The only thing that writes ->next is this

```
		ftp = lws_malloc(sizeof(*ftp), "ftp");
		...
		ftp->next = NULL;

		/* place at END of list to maintain order */
		ftp1 = &pt->foreign_pfd_list;
		while (*ftp1)
			ftp1 = &(*ftp1)->next;

		*ftp1 = ftp;
```

> Note that If I clear the list pointer before handling the different queued events, the issue disappears:

Is there maybe some problem with volatile members with that compiler?  It should be that you cannot be executing the code that creates the ftps at the time we are draining it.

Can you maybe add some logging and confirm what sequence of events makes that happen?

**Comment by lws-team** [comment]
Ahhh are you calling the lws api from multiple threads without any locking on your side?

**Comment by RouquinBlanc** [comment]
No the code is just calling on_writable again and again from a single
thread... I suspect that during the wiping of the queue, it is still
written by someone else in parallel. Will investigate a bit more later!

**Comment by lws-team** [comment]
I tried your test app, it can compile on Linux almost the same

```
c++ -std=c++14 -O0 -g -I/usr/local/include -I/usr/include/openssl -I/usr/local/include -L/usr/local/lib  -lwebsockets -lz  -o wsdemo main.cpp /usr/local/lib/libwebsockets.a -pthread
```

I adapted it to offer `lws-mirror-protocol` so I can connect to it.  After some time it spins at 100% as you say.  So I wil also study it.

**Comment by lws-team** [comment]
Okie... thanks to your test app and help, this is solved now I think.  Please retry with latest master.

The solution involves using volatile and memory barriers to make sure the threads see each others' true situation.   With that your test app can run for as long as you like with continuous WRITEABLE callbacks coming on the service thread.

**Comment by RouquinBlanc** [comment]
OK, working as well on my side.

Now let's get to the next test I wanted to make: multiple sockets handled at the same time.

Look at [this script](https://pastebin.com/YG705z4X): It spawns a new client thread each time a ws connects and just keeps calling on_writable. I've been connecting like crazy as many sockets as I could in parallel and it hangs together perfectly 👍 👍 👍 

Note that if I don't wrap external calls to `lws_callback_on_writable` under a global mutex (same for all clients), it crashes miserably on a segmentation fault on the `while (*ftp1)` in pollfd.c:113.
Just comment out the lock/unlock in `client_loop` to see it happen with just 2 parallel clients.

Should the external calls be protected directly inside the call? After all, you know that the call is done from outside the service thread, it would be easy to have an "external context client lock" mechanism for authorised calls from the outside like this one, don't you think? Less error prone.

Thanks for the great work anyway!

**Comment by lws-team** [comment]
I'm glad it's working as far as it goes.

All the struggle with that today was only about making bsd poll() act like linux poll() already did.. it doesn't help with libuv, libev, or libevent event loop backends lws also supports, which are completely singlethreaded and lockless, being suddenly threadsafe.  Their implementation is opaque so I can't bracket their entry and exit from their wait as I did poll() either, they just can't use this trick.

I think it's ok putting the serialization burden on the caller... not only because then we don't have to match whatever threading lib they usedinside lws... at the moment for poll() backend it's as you saw today, should be robust against one foreign thread calling on_writable at a time (it's not hard to write a 10-line wrapper with their own locking to hide the serialization restriction); for other backends they should use cancel service api, and do their own locking in the serve thread context at LWS_CALLBACK_EVENT_WAIT_CANCELLED to walk an external list of guys who need on_writable() calling.

**Comment by RouquinBlanc** [comment]
Fair enough. Then I think it would be worth mentioning that in the documentation of `lws_callback_on_writable` somewhere, just in case!

Thanks for fixing this, I'm going to go back at fixing janus now ;-)

**Comment by lws-team** [comment]
Yes... I rewrote the section in README.coding.md to give a better overview of the situation and the reasons things are like that and updated the patch.  Thanks again for your help.
